### ISSUE TITLE
Bug: [no-unnecessary-type-assertion] false positive when casting IIFE with `void` return type to `undefined`

### ISSUE THREAD

**Issue by szhsin** [typescript-eslint/typescript-eslint#11529]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.9.2&fileType=.ts&code=PTAEhlyUDMEMBsDOBTUAHA9gglgFywNyQCgBjdAOwR1AFUA5AEQFEAxASTqYdAF5QAKfgEpeAPlABvUAF8hw0DASgAruQAmSKFnJI1AbiJEQoQKDkoOulBIATtfTXSFKtF4CRPcVOkGylavWZ2TgYAJlcoeUUVdU1tXQMgA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0ipWkOTJE0fJQ5N0UOdA7RI4MAF8QOoA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
// ❌ false positive
const UNDEFINED = (() => { })() as undefined;

// ✅ No error
const f = () => { };
const UNDEFINED2 = f() as undefined;
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-unnecessary-type-assertion": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No error

### Actual Result

Error: this assertion is unnecessary since it does not change the type of the expression.

### Additional Info

This might be a minor issue, but it could indicate a problem with the rule. Casting an IIFE with a `void` return type to `undefined` triggers a false positive.

**Comment by bradzacher** [comment]
This is technically correct because the `void` type does not strictly mean `undefined` it means `any`.

**Comment by kawaaaas** [comment]
Hi @bradzacher,

I frequently use this great tool and am grateful to all the maintainers and contributors.

As you correctly pointed out, this IIFE is technically evaluated as undefined by TypeScript's type checker:

```typescript
console.log(typeof (() => {})());
// TypeScript infers this as undefined
```

So my understanding is that while TypeScript's type checker identifies this as undefined, the intent here is to allow this type assertion pattern in the ESLint rule, correct?
If so, I'd like to modify the rule so that ESLint doesn't report an error when it encounters an IIFE that TypeScript interprets as undefined.

Would you be open to assigning this issue to me if my understanding is correct? Please let me know if I'm misunderstanding anything about the intended behavior.
